pkg/handler/middleware/middleware.go

```diff
@@ -444,3 +448,46 @@ func getConstraintProvider(clusterProviderGetter provider.ClusterProviderGetter,
 
 	return constraintProviderGetter(seed)
 }
+
+// Alertmanagers is a middleware that injects the current AlertmanagerProvider into the ctx
+func Alertmanagers(clusterProviderGetter provider.ClusterProviderGetter, alertmanagerProviderGetter provider.AlertmanagerProviderGetter, seedsGetter provider.SeedsGetter) endpoint.Middleware {
+	return func(next endpoint.Endpoint) endpoint.Endpoint {
+		return func(ctx context.Context, request interface{}) (response interface{}, err error) {
+			seedCluster := request.(seedClusterGetter).GetSeedCluster()
+
+			alertmanagerProvider, err := getAlertmanagerProvider(clusterProviderGetter, alertmanagerProviderGetter, seedsGetter, seedCluster.SeedName, seedCluster.ClusterID)
+			if err != nil {
+				return nil, err
+			}
+			ctx = context.WithValue(ctx, AlertmanagerProviderContextKey, alertmanagerProvider)
+			return next(ctx, request)
+		}
+	}
+}
+
+func getAlertmanagerProvider(clusterProviderGetter provider.ClusterProviderGetter, alertmanagerProviderGetter provider.AlertmanagerProviderGetter, seedsGetter provider.SeedsGetter, seedName, clusterID string) (provider.AlertmanagerProvider, error) {
+	seeds, err := seedsGetter()
+	if err != nil {
+		return nil, err
+	}
+
+	if clusterID != "" {
+		for _, seed := range seeds {
+			clusterProvider, err := clusterProviderGetter(seed)
+			if err != nil {
```

Comment:
What if the error wasn't 404?

Reply:
Actually, I didn't look into this in details, just copied from `getAddonProvider` now I think it makes more sense to return something like `KubernetesErrorToHTTPError` https://github.com/kubermatic/kubermatic/blob/master/pkg/handler/v1/common/error.go#L27
WDYT?

Reply:
Yes, `KubernetesErrorToHTTPError` is better

Reply:
Done